### PR TITLE
Adding toggle system for old/new/all trains

### PR DESCRIPTION
**Pull request by devinmatte**
# Motivation

As we move towards some lines being almost entirely new trains (orange), we should be able to see how many old trains are running as well

# Changes
- Adds a new toggle to switch between old/new/all
- Has the API return three lists (old, new and all)
    - we can totally do the filtering of the list on the frontend to reduce network overhead, this was just my first attempt
- Cleans up the Line selector to Typescript and adds a series of types to make life easier

https://user-images.githubusercontent.com/9310513/184235852-59e84b1d-9ebd-4d02-b9fc-f413329b9f80.mov

### CODE REVIEW

File: src/components/AgeTabPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    { key: 'new_vehicles', label: 'New' },
+    { key: 'vehicles', label: 'All' },
+];
+
+interface AgeTabPickerProps {
+    tabState: TabStateReturn;
+    tabColor: string;
+}
+
+export const AgeTabPicker: React.FC<AgeTabPickerProps> = ({ tabState, tabColor }) => {
```
Comment (idreyn) [inline src/components/AgeTabPicker.tsx:20]: This component duplicates a fair amount of the logic in the existing `TabPicker`. Could we make that component a bit more generic and then reuse it for a `LineTabPicker` and an `AgeTabPicker`?

File: src/main.css [hunk 2]
```diff
@@ -186,6 +194,11 @@ a {
     font-weight: bold;
 }
 
+.tab-picker > .tab > .age {
```
Comment (idreyn) [inline src/main.css:197]: I wonder if this should actually be even smaller. It looks a little cramped with just the text NEW in a bubble.

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+export interface Line {
```
Comment (idreyn) [inline src/types.ts:1]: Oh, this is really lovely, thanks.

**Review by idreyn — APPROVED** [review]
This looks really nice! I have a few suggestions for cleanup:

- We could save a few bytes by unconditionally sending down _all_ trains from the server and doing the new/old/all filtering in `useMbtaApi()`.
- The `AgeTabPicker` and the existing `TabPicker` for lines could both use a shared component with the underlying logic.

But I also think this is a worthy improvement to the NTT as-is, so take those as suggestions.

File: src/components/AgeTabPicker.tsx (hunk 1 above)
Comment (devinmatte) [reply src/components/AgeTabPicker.tsx:20]: They share some logic, but now the `useLayoutEffect`s are different, as well as how it generates the tablist. We could extract out some logic, but I think we'd end up with two files with similar lengths and complexities. I might say put a pin in it (make an issue) if we care about making it more generic.

File: src/main.css (hunk 2 above)
Comment (devinmatte) [reply src/main.css:197]: Yeah agreed. I'll try a little more messing around to see what looks good

File: src/components/util.ts [hunk 4]
```diff
@@ -1,9 +1,10 @@
-import React from 'react';
+import { createContext } from 'react';
+import { Route, Routes, Train } from '../types';
 
-export const PopoverContainerContext = React.createContext(null);
+export const PopoverContainerContext = createContext(null);
 
-export const getTrainRoutePairsForLine = (trainsByRoute, routes) => {
-    const pairs = [];
+export const getTrainRoutePairsForLine = (trainsByRoute: { [key: string]: Train[]}, routes: Routes) => {
```
[the remark this answers is not in the document]
Comment (devinmatte) [reply src/components/util.ts]: Tbh I forgot about `Record`, that's exactly what I was looking for

**Comment by devinmatte** [comment]
Screen Shot 2022-08-08 at 6 34 29 PM
Okay I think the pills look a little less cramped now

**Review by nathan-weinberg — APPROVED** [review]
LGTM - I wonder if we might want to put the new trains tab in front of the old since that's the focus? Also wondering if we might want to make a change to the tagline for the app. Anyway both nits won't block on either.

Thanks for doing this @devinmatte! I think this definitely warrants a version bump to `2.2.0`

**Comment by devinmatte** [comment]
@nathan-weinberg Happy to update the tagline, do we want to just cut `new` or do we maybe want to update it based on the selection? Whatever the consensus is I'm happy to change it to
As for the `new` being first, I think it looks better centered. So we're selecting the middle option to start, it's the focus still.

**Comment by nathan-weinberg** [comment]
> @nathan-weinberg Happy to update the tagline, do we want to just cut `new` or do we maybe want to update it based on the selection? Whatever the consensus is I'm happy to change it to

I kind of like both ideas - let's see what others have to say

> As for the `new` being first, I think it looks better centered. So we're selecting the middle option to start, it's the focus still.

Fair enough

**Pull request merged**